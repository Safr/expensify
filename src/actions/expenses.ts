import { Dispatch, ActionCreator } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { toastr } from 'react-redux-toastr';
import database from '../firebase/firebase';
import { SET_EXPENSES, ADD_EXPENSE, EDIT_EXPENSE, REMOVE_EXPENSE } from '../actionTypes';
import { asyncActionStart, asyncActionFinish, asyncActionError } from './async';
import StoreState, { ExpensesI } from '../types';

export interface InitExpense {
  type: '@@INIT';
}

export interface AddExpense {
  type: ADD_EXPENSE;
  expense: ExpensesI;
}

export interface EditExpense {
  type: EDIT_EXPENSE;
  id: string;
  updates: any;
}

export interface RemoveExpense {
  type: REMOVE_EXPENSE;
  id: string;
}

export interface SetExpenses {
  type: SET_EXPENSES;
  expenses: ExpensesI[];
}

export type ExpenseAction = InitExpense | AddExpense | EditExpense | RemoveExpense | SetExpenses;

export const addExpense: ActionCreator<AddExpense> = expense => ({
  type: ADD_EXPENSE,
  expense,
});

export const startAddExpense = (expenseData = {}): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0,
  }: any = expenseData;
    const expense = {
      description, note, amount, createdAt,
    };
    return database.ref(`users/${uid}/expenses`).push(expense).then(ref => {
      toastr.success('Success', 'Expense has been added');
      dispatch(addExpense({
        id: ref.key,
        ...expense,
      }));
    });
  };

export const removeExpense: ActionCreator<RemoveExpense> = (id) => ({
  type: REMOVE_EXPENSE,
  id,
});

export const startRemoveExpense = (id: string): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    return database.ref(`users/${uid}/expenses/${id}`).remove()
    .then(() => {
      toastr.success('Success', 'Expense has been deleted');
      dispatch(removeExpense(id));
    });
  };

export const editExpense: ActionCreator<EditExpense> = (id, updates) => ({
  type: EDIT_EXPENSE,
  id,
  updates,
});

export const startEditExpense = (id: string, updates: ExpensesI): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates)
    .then(() => {
      toastr.success('Success', 'Expense has been edited');
      dispatch(editExpense(id, updates));
    });
  };

export const setExpenses: ActionCreator<SetExpenses> = expenses => ({
  type: SET_EXPENSES,
  expenses,
});

export const startSetExpenses = (): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    dispatch(asyncActionStart());
    return database.ref(`users/${uid}/expenses`)
    .once('value')
    .then(snapshot => {
      const expenses: ExpensesI[] = [];
      snapshot.forEach((childSnapshot: any) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val(),
        });
      });
      dispatch(setExpenses(expenses));
      toastr.success('Success', 'Expenses have been loaded');
      dispatch(asyncActionFinish());
    })
    .catch(e => dispatch(asyncActionError()));
  };
