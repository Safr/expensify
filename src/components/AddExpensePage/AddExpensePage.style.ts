import styled from 'styled-components';
import { blue, fontSizeLarge, fontSizeSmall, lSize, mSize, offWhite, sSize, white } from '../../theme/variables';
import media from '../../theme/media';

const PageHeader = styled.div`
  padding: ${sSize};
  font-size: ${fontSizeSmall};
  background-color: ${offWhite};
  ${media.tablet`
    padding: ${lSize};
    font-size: ${fontSizeLarge};
  `}

  h2 {
    text-align: left;
  }
`;

export default PageHeader;
