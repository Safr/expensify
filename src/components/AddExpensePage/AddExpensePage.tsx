import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import ExpenseForm from '../ExpenseForm/ExpenseForm';
import { startAddExpense } from '../../actions';
import { ExpensesI, RouteComponentProps } from '../../types';
import PageHeader from './AddExpensePage.style';

interface Props extends RouteComponentProps<any> {
  onStartAddExpense: (expense: ExpensesI) => void;
}

export class AddExpensePage extends Component<Props> {
  handleSubmit = (expense: any) => {
    this.props.onStartAddExpense(expense);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <PageHeader>
          <h1>Add Expense</h1>
        </PageHeader>
        <ExpenseForm onSubmit={this.handleSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
  onStartAddExpense: (expense: ExpensesI) => dispatch(startAddExpense(expense)),
});

export default connect(null, mapDispatchToProps)(AddExpensePage);
