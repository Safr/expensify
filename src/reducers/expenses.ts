import { SET_EXPENSES, ADD_EXPENSE, EDIT_EXPENSE, REMOVE_EXPENSE } from '../actionTypes';
import { ExpenseAction } from '../actions';
import { ExpensesI } from '../types';

const expensesReducer = (state = [], action: ExpenseAction): ExpensesI | ExpensesI[] => {
  switch (action.type) {
    case SET_EXPENSES:
      return [...action.expenses];
    case ADD_EXPENSE:
      return [
        ...state,
        action.expense,
      ];
    case REMOVE_EXPENSE:
      return state.filter(({ id }: {id: string}) => id !== action.id);
    case EDIT_EXPENSE:
      return state.map(expense => (
        expense.id === action.id ? { ...expense, ...action.updates } : expense));
    default:
      return state;
  }
};

export default expensesReducer;

