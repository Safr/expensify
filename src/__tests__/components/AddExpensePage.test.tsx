import React from 'react';
import { AddExpensePage } from '../../components/AddExpensePage/AddExpensePage';
import expenses from '../../fixtures/expenses';
import { shallow } from 'enzyme';

describe('AddExpensePage component', () => {
  let onStartAddExpense;
  let history;
  let wrapper;
  beforeEach(() => {
    onStartAddExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage onStartAddExpense={onStartAddExpense} history={history} />);
  });
  test('should render AddExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  test('should handle onStartAddExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onStartAddExpense).toHaveBeenLastCalledWith(expenses[0]);
  });
});
