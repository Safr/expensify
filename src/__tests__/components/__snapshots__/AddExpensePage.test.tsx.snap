// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddExpensePage component should render AddExpensePage correctly 1`] = `
<div>
  <styled.div>
    <h1>
      Add Expense
    </h1>
  </styled.div>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
